rfcs/proposed/dynamic_selection_customization/README.md

```diff
@@ -0,0 +1,436 @@
+        using resource_type = ResourceType;
+
+        // space needed by the backend in the Selection type
+        // for implementing reporting, by default an empty struct
+        template<typename ...Info> struct scratch_t;
+
+        // constructors
+        backend_base() {}
+        backend_base(const std::vector<resource_type>& u) : resources_{v} {}
+
+        template <typename SelectionHandle, typename Function, typename... Args>
+        auto submit(SelectionHandle s, Function&& f, Args&&... args) {
+            return static_cast<Backend*>(this)->submit_impl(s, f, std::forward<Args>(args)...);
+        }
+
+        std::vector<resource_type> get_resources() {
+            return static_cast<Backend*>(this)->get_resources_impl();
+        }
+
+        auto get_submission_group() {
+            return static_cast<Backend*>(this)->get_submission_group();
+        }
+
+        template <typename SelectionHandle, typename Function, typename... Args>
+        auto submit(SelectionHandle s, Function&& f, Args&&... args) {
+            return static_cast<Backend*>(this)->submit_impl(s, f, std::forward<Args>(args)...);
+        }
+
+    protected:
+
+        //
+        // The default functionality that can be overriden in derived classes
+        //
+
+        std::vector<resource_type> resources_;
+
+        template <typename SelectionHandle, typename Function, typename... Args>
+        auto submit_impl(SelectionHandle s, Function&& f, Args&&... args) {
+        {
+            static_cast<Backend*>(this)->instrument_before_impl(s);
+            return static_cast<Backend*>(this)->instrument_after_impl(
+                s, 
+                std::forward<Function>(f)(oneapi::dpl::experimental::unwrap(s), 
+                                          std::forward<Args>(args)...)
+            );
+        }
+
+        template <typename SelectionHandle>
+        void instrument_before_impl(SelectionHandle s) { }
+
+        template <typename SelectionHandle, typename WaitType>
+        auto instrument_after_impl(SelectionHandle s, WaitType w) {
+            return default_submission{w};
+        }
+
+        std::vector<resource_type> get_resources_impl() const noexcept {
+            return resources_;
+        }
+
+    };
+```
+
+We also propose to provide a catch-all [default backend](#default-backend) for uncustomized resource
+types:
+
+```cpp
+    // A general default backend for ResourceType
+    template< typename ResourceType >
+    class default_backend : public backend_base<ResourceType, default_backend<ResourceType>> {
+    public:
+        using resource_type = ResourceType;
+    };
+```
+
+A backend that relies on the defaults but overrides some functionality can then specialize the
+default backend for the type. For example, the existing SYCL backend can be implemented as:
+
+```cpp
+template< >
+class default_backend<sycl::queue> : public backend_base<sycl::queue, default_backend<sycl::queue>>
+{
+  //* override as needed
```

Comment:
Is this a TODO for the RFC?
Do we want to describe what is actually necessary to define `sycl::queue` backend here?


In the PoC as it stands, it actually its ~30 lines longer implementation than the previous version... can we cut it down more?

Or is there a different simpler example we should use here?

Reply:
I will add different examples.